### DISCUSSION TITLE
Help needed with clone private GitLab repo with contracts

### DISCUSSION THREAD

**Discussion by tkrisztian95** [specmatic/specmatic#900] [Ask For Help] [answered]
Hi,

I have a `specmatic.json` at the consumer project side. The config json at `sources[].repository` specify the Git repo url and it contains a GitLab CI job token for accessing the stub yaml file from a private repo.

As its an option documented here https://specmatic.in/documentation/specmatic_json.html#source-control-authentication
> Alternatives - With https URLs you can provide the bearer token or other means. Please reach us (raise a [github issue](https://github.com/znsio/specmatic/issues/new/choose)) if you need help with this.

I would like to avoid the token be pushed to the git in this config file.

Is it possible to use environment variable in the repository url and specmatic resolve it from the CI scope? Or is there other ways to pass this token?

PS: Unfortunately the token can be also be found in the `stub_usage_report.json` created in the build output when running the test. So it can be a blocker for using the token in the git url. We need to find another way to allow clone the private repo with the contracts.

**Update**
We are considering to use [Deploy keys](https://docs.gitlab.com/ee/user/project/deploy_keys/index.html) to allow clone the repository with the contracts.

Thanks!

**Comment by tkrisztian95** [comment]
Any recommendations from the Specmatic team, how to get this done properly on GitLab?

**Comment by tkrisztian95** [comment, marked as answer]
Configuring Deploy keys worked on GitLab.

**Reply by harikrishnan83** [reply]
@tkrisztian95 that is great. Thanks for sharing this.

**Comment by harikrishnan83** [comment]
@tkrisztian95 Specmatic now substitutes env variables that are part of the Git repo URL that you mention in specmatic.json. This is part of the [latest release 1.3.18](https://github.com/znsio/specmatic/releases/tag/1.3.18). Here is the [link to the documentation](https://specmatic.in/documentation/continuous_integration.html#authentication-params-in-git-repo-uri). With this you should be able to use Gitlab env variables as part of the Repo URI. Examples:

* CI_JOB_TOKEN - `https://gitlab-ci-token:${CI_JOB_TOKEN}@gitlab.com/contract-testing/central-contract-repo.git/`
* Username / Password - `https://${USERNAME}:${PASSWORD}@gitlab.com/contract-testing/central-contract-repo.git/`

Adding this in this thread in addition to your deploy keys based solution for anyone referring to this discussion.
